File: cpp/2418.sort-the-people.cpp
#include <algorithm>
#include <string>
#include <utility>
#include <vector>

using std::pair;
using std::sort;
using std::string;
using std::vector;

class Solution {
public:
  vector<string> sortPeople(vector<string> &names, vector<int> &heights) {
    vector<pair<int, int>> tmp;
    for (int i = 0; i < heights.size(); ++i) {
      tmp.push_back({heights[i], i});
    }

    sort(tmp.begin(), tmp.end());

    vector<string> ret;
    for (int i = tmp.size() - 1; i >= 0; --i) {
      ret.push_back(names[tmp[i].second]);
    }
    return ret;
  }
};

File: golang/2418.sort-the-people.go
type Person struct {
	Name   string
	Height int
}

type ByHeight []Person

func (b ByHeight) Len() int {
	return len(b)
}

func (b ByHeight) Swap(i, j int) {
	b[i], b[j] = b[j], b[i]
}

func (b ByHeight) Less(i, j int) bool {
	return b[i].Height > b[j].Height
}

func sortPeople(names []string, heights []int) []string {
	tmp := make([]Person, 0, len(heights))
	for i := 0; i < len(heights); i++ {
		tmp = append(tmp, Person{Name: names[i], Height: heights[i]})
	}

	sort.Sort(ByHeight(tmp))

	ret := make([]string, len(tmp))
	for i := 0; i < len(tmp); i++ {
		ret[i] = tmp[i].Name
	}
	return ret
}
